src/Castle.Core/DynamicProxy/ProxyUtil.cs

```diff
@@ -109,17 +150,134 @@ public static bool IsAccessible(MethodBase method, out string message)
 			return false;
 		}
 
+		/// <summary>
+		/// Checks whether the specified type is accessible to DynamicProxy.</summary>
+		/// <param name="type">The type to check.</param>
+		/// <returns><c>true</c> if the type is accessible to DynamicProxy, <c>false</c> otherwise.</returns>
+		public static bool IsAccessible(Type type)
+		{
+			return IsAccessibleType(type);
+		}
+
+		internal static bool IsAccessibleType(Type target)
+		{
+			var typeInfo = target.GetTypeInfo();
+
+			var isPublic = typeInfo.IsPublic || typeInfo.IsNestedPublic;
+			if (isPublic)
+			{
+				return true;
+			}
+
+			var isTargetNested = target.IsNested;
+			var isNestedAndInternal = isTargetNested && (typeInfo.IsNestedAssembly || typeInfo.IsNestedFamORAssem);
+			var isInternalNotNested = typeInfo.IsVisible == false && isTargetNested == false;
+			var isInternal = isInternalNotNested || isNestedAndInternal;
+			if (isInternal && AreInternalsVisibleToDynamicProxy(typeInfo.Assembly))
+			{
+				return true;
+			}
+
+			return false;
+		}
+
+		/// <summary>
+		///   Checks whether the specified method is accessible to DynamicProxy.
+		///   Unlike with <see cref="IsAccessible(MethodBase)"/>, the declaring type's accessibility is ignored.
+		/// </summary>
+		/// <param name = "method">The method to check.</param>
+		/// <returns><c>true</c> if the method is accessible to DynamicProxy, <c>false</c> otherwise.</returns>
+		internal static bool IsAccessibleMethod(MethodBase method)
+		{
+			// Accessibility supported by the full framework and CoreCLR
+			if (method.IsPublic || method.IsFamily || method.IsFamilyOrAssembly)
+			{
+				return true;
+			}
+
+			if (method.IsFamilyAndAssembly)
+			{
+				return true;
+			}
+
+			if (AreInternalsVisibleToDynamicProxy(method.DeclaringType.GetTypeInfo().Assembly) && method.IsAssembly)
+			{
+				return true;
+			}
+
+			return false;
+		}
+
+		/// <summary>
+		///   Determines whether the specified method is internal.
+		/// </summary>
+		/// <param name = "method">The method.</param>
+		/// <returns>
+		///   <c>true</c> if the specified method is internal; otherwise, <c>false</c>.
+		/// </returns>
+		internal static bool IsInternal(MethodBase method)
+		{
+			return method.IsAssembly || (method.IsFamilyAndAssembly && !method.IsFamilyOrAssembly);
+		}
+
+		/// <summary>
+		/// Provides instructions that a user could follow to make a type or method in <paramref name="targetAssembly"/>
+		/// visible to DynamicProxyGenAssembly2.</summary>
+		/// <param name="targetAssembly">The assembly containing the type or method.</param>
+		/// <returns>Instructions that a user could follow to make a type or method visible to DynamicProxyGenAssembly2.</returns>
+		internal static string CreateInstructionsToMakeVisible(Assembly targetAssembly)
```

Comment:
I think this method should go into `ExceptionMessageBuilder` instead, otherwise `ProxyUtil` is going to get big quickly.

The other option would be to keep `InternalsUtil` which can hold the implementation and `ProxyUtil` just forwards through, we could keep the obsolete public methods at the top of the file for now forwarding to internal methods below. Thoughts?